File: src/ts/magma/app/compile/type/collect/variadic/VariadicTypes.ts
import { CompileState } from "../../../../../../magma/app/compile/CompileState";
import { Type } from "../../../../../../magma/app/compile/type/Type";
import { Tuple2 } from "../../../../../../magma/api/Tuple2";
import { Option } from "../../../../../../magma/api/option/Option";
import { Strings } from "../../../../../../magma/api/text/Strings";
import { SuffixComposable } from "../../../../../../magma/app/compile/compose/SuffixComposable";
import { TypeCompiler } from "../../../../../../magma/app/TypeCompiler";
import { Tuple2Impl } from "../../../../../../magma/api/Tuple2Impl";
import { VariadicType } from "../../../../../../magma/app/compile/type/collect/variadic/VariadicType";
export class VariadicTypes {
	static parseVariadic(state: CompileState, input: string): Option<Tuple2<CompileState, Type>> {
		let stripped = Strings.strip(input)/*unknown*/;
		return new SuffixComposable<Tuple2<CompileState, Type>>("...", (s: string) => {
			return TypeCompiler.createTypeRule().apply(state, s).map((tuple: Tuple2<CompileState, Type>) => new Tuple2Impl<CompileState, Type>(tuple.left(), tuple.right())/*unknown*/).map((child: Tuple2Impl<CompileState, Type>) => new Tuple2Impl<CompileState, Type>(child.left(), new VariadicType(child.right()))/*unknown*/)/*unknown*/;
		}).apply(stripped)/*unknown*/;
	}
	static generateVariadicType(variadicType: VariadicType): string {
		return TypeCompiler.generateType(variadicType.type()) + "[]"/*unknown*/;
	}
}

File: src/java/magma/app/compile/type/collect/variadic/VariadicTypes.java
package magma.app.compile.type.collect.variadic;

import magma.api.Tuple2;
import magma.api.Tuple2Impl;
import magma.api.option.Option;
import magma.api.text.Strings;
import magma.app.TypeCompiler;
import magma.app.compile.CompileState;
import magma.app.compile.compose.SuffixComposable;
import magma.app.compile.type.Type;

public final class VariadicTypes {
    public static Option<Tuple2<CompileState, Type>> parseVariadic(CompileState state, String input) {
        var stripped = Strings.strip(input);
        return new SuffixComposable<Tuple2<CompileState, Type>>("...", (String s) -> {
            return TypeCompiler.createTypeRule().apply(state, s)
                    .map((Tuple2<CompileState, Type> tuple) -> new Tuple2Impl<CompileState, Type>(tuple.left(), tuple.right()))
                    .map((Tuple2Impl<CompileState, Type> child) -> new Tuple2Impl<CompileState, Type>(child.left(), new VariadicType(child.right())));
        }).apply(stripped);
    }

    public static String generateVariadicType(VariadicType variadicType) {
        return TypeCompiler.generateType(variadicType.type()) + "[]";
    }
}
